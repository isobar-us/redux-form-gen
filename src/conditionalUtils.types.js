import type {FieldType} from './types';
export type ConditionalObject = {
  questionId: string,
  equals?: mixed,
  // cond
  and?: Array<ConditionalObject>,
  or?: Array<ConditionalObject>,
  not?: ConditionalObject,
  // lodash
  filled?: boolean,
  includes?: mixed | Array<mixed>,
  // comparison
  greaterThan?: number,
  lessThan?: number,
  greaterThanEqual?: number,
  lessThanEqual?: number,
  // length
  length: number,
  minLength: number,
  maxLength: number,
  // regex
  pobox: boolean,
  email: boolean,
  regex: boolean
};

export type ConditionalDependency = {
  questionId: string,
  globalScope?: Boolean
};

export type FieldDependencyOptions = {
  field: FieldType,
  parentPath: String
};

export type EvalCondOptions = {
  elseHandler?: Function,
  valueKey?: string,
  cond: ConditionalObject,
  customOperators: CustomOperators // eslint-disable-line no-use-before-define
};

export type CustomOperators = {
  [key: string]: {(options: EvalCondOptions): boolean}
};

export type ConditionalOperatorOptions = {
  value: mixed,
  param: mixed,
  ...EvalCondOptions
};

type ConditionalOperatorNumberParam = {
  (options: {...ConditionalOperatorOptions, value: number, param: number}): boolean
};
type ConditionalOperatorObjectParam = {
  (options: {...ConditionalOperatorOptions, value: Object, param: boolean}): boolean
};

type ConditionalOperator = {(options: ConditionalOperatorOptions): boolean};

export type ConditionalOperators = {
  equals: ConditionalOperator,
  and: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  or: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  not: {(options: {...ConditionalOperatorOptions, param: ConditionalObject}): boolean},
  filled: ConditionalOperator,
  includes: {
    (options: {...ConditionalOperatorOptions, value: Array<mixed>, param: mixed}): boolean
  },
  greaterThan: ConditionalOperatorNumberParam,
  lessThan: ConditionalOperatorNumberParam,
  greaterThanEqual: ConditionalOperatorNumberParam,
  lessThanEqual: ConditionalOperatorNumberParam,
  length: ConditionalOperatorObjectParam,

  gt?: ConditionalOperatorNumberParam,
  lt?: ConditionalOperatorNumberParam,
  min?: ConditionalOperatorNumberParam,
  max?: ConditionalOperatorNumberParam,
  value?: ConditionalOperator
};
