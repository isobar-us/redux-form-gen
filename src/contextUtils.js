// @flow
import React from 'react';
import PropTypes from 'prop-types';

import createReactContext from 'create-react-context';
export const GenContext = createReactContext({wasGenerated: false});

export const consumeGenContext = (Component: Function) => (props: mixed) => (
  <GenContext.Consumer>
    {(gen) => <Component {...props} gen={gen} /> }
  </GenContext.Consumer>
);

// inspired by
// https://medium.com/react-ecosystem/how-to-handle-react-context-a7592dfdcbc

// const provideContext = (childContextTypes, getChildContext) => (WrappedComponent) => {
//   class ContextProvider extends Component {
//     static childContextTypes = childContextTypes;
//     getChildContext = () => getChildContext(this.props);
//
//     render() {
//       return <WrappedComponent {...this.props} />;
//     }
//   }
//   return ContextProvider;
// };

const consumeContext = (contextTypes) => (Component: Function) => {
  /* The context is passed as props. This way the component is
   completely decoupled from the context API.
  */
  const ContextConsumer = (props: mixed, context: mixed) => <Component {...props} {...context} />;
  ContextConsumer.contextTypes = contextTypes;
  return ContextConsumer;
};

export const reduxFormContext = {
  _reduxForm: PropTypes.object
};

export const consumeReduxFormContext = consumeContext(reduxFormContext);
