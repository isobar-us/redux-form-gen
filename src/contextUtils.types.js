import type {Props as FormGeneratorProps} from './FormGenerator.types';

export type GenContextProps = {
  ...FormGeneratorProps,
  getCachedValue: Function,
  setCachedValue: Function
};
