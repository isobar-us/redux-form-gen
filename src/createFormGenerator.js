import React from 'react';
import {
  FormGenerator,
  getDefaultValues,
  getFieldOptions,
  buildLookupTable,
  getSectionErrors,
  evalCond,
  injectGenProps,
  isFieldFilled,
  isFieldEmpty,
  isFieldVisible,
  mapFieldChildren,
  isNilOrEmpty
} from '@isobar-us/redux-form-gen';

const mergeCustomOptions = (customOptions, options) => ({
  ...options,
  customFieldTypes: {...customOptions.customFieldTypes, ...(options.customFieldTypes || {})},
  customOperators: {...customOptions.customOperators, ...(options.customOperators || {})}
});

const wrapComponent = (Wrapped, customOptions) => {
  const ComponentWrapper = (options) => <Wrapped {...mergeCustomOptions(customOptions, options)} />;

  return ComponentWrapper;
};

const wrapFunction = (wrapped, customOptions) => {
  return (options, ...args) => wrapped(mergeCustomOptions(customOptions, options), ...args);
};

const createFormGenerator = (customOptions) => {
  return {
    // functions
    getDefaultValues: wrapFunction(getDefaultValues, customOptions),
    getFieldOptions: wrapFunction(getFieldOptions, customOptions),
    buildLookupTable: wrapFunction(buildLookupTable, customOptions),
    getSectionErrors: wrapFunction(getSectionErrors, customOptions),
    evalCond: wrapFunction(evalCond, customOptions),
    isFieldFilled: wrapFunction(isFieldFilled, customOptions),
    isFieldEmpty: wrapFunction(isFieldEmpty, customOptions),
    isFieldVisible: wrapFunction(isFieldVisible, customOptions),
    mapFieldChildren: wrapFunction(mapFieldChildren, customOptions),
    // components
    FormGenerator: wrapComponent(FormGenerator, customOptions),
    injectGenProps: (Form) => {
      const GenForm = injectGenProps(Form);
      const InjectedGenProps = (options) => <GenForm {...mergeCustomOptions(options)} />;
      return InjectedGenProps;
    },
    // non-wrapped
    isNilOrEmpty
  };
};

export default createFormGenerator;
