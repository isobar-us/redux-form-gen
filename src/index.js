// @flow
import FormGenerator from './FormGenerator';

import defaultFieldTypes, {getFieldOptions, genericFieldProps} from './defaultFieldTypes';
import GenericRequiredLabel from './defaultFieldTypes/components/GenericRequiredLabel';
import RequiredIndicator from './defaultFieldTypes/components/RequiredIndicator';
import injectGenProps from './injectGenProps';

import {
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren
} from './validators';

import {getDefaultValues, buildLookupTable} from './utils';

import {consumeGenContext, GenContext} from './contextUtils';

import {evalCond, evalCondValid} from './conditionalUtils';

import GenField from './GenField';
import GenWrapper from './GenWrapper';

export type {FieldOptions, CustomFieldTypes} from './types';

export {
  FormGenerator,
  // defaultFieldTypes
  defaultFieldTypes,
  getFieldOptions,
  RequiredIndicator,
  GenericRequiredLabel,
  genericFieldProps,
  injectGenProps,
  // validators
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  // validator utils
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren,
  // utils
  getDefaultValues,
  buildLookupTable,
  // contextUtils
  consumeGenContext,
  GenContext,
  // conditionalUtils
  evalCond,
  evalCondValid,
  // internals
  GenField,
  GenWrapper
};
