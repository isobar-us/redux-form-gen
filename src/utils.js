// @flow
import set from 'lodash/set';
import has from 'lodash/has';
import isNil from 'lodash/isNil';
import {getFieldOptions} from './defaultFieldTypes';
import defaultsDeep from 'lodash/defaultsDeep';
import {isFieldHidden, getFieldPath, resolveFieldOptions, mapFieldChildren} from './validators';

import type {BuildLookupTableOptions, GetDefaultValuesOptions, GetDefaultValueOptions} from './utils.types';

export const buildLookupTable = (options: BuildLookupTableOptions, table: Object = {}) => {
  const {fields, customFieldTypes} = options;
  if (fields) {
    fields.map((field) => {
      const fieldOptions = getFieldOptions({field, customFieldTypes});

      /* TODO use questionRef for unique ids?
         - this would allow for multiple fields to reference the same questionId
         - would need to transform questionRef => questionId when running evalCond using lookupTable?
      */

      if (!isNil(fieldOptions)) {
        // only parse valid field types
        if (field.questionId) {
          if (!isNil(table[field.questionId])) {
            // console.error('duplicate question id', field.questionId, field);
          } else {
            // if (Array.isArray(field.fields)) {
            //   set(table, field.questionId, newBuildLookupTable({
            //     ...options,
            //     fields: field.fields
            //   }));
            // } else {
            // console.info('setting field', field.questionId);
            set(table, field.questionId, field);
            // }
          }
        }

        // const table = {
        //   questionId: {
        //     item: {...field}
        //     childField: {
        //       subQuestionId: {
        //         ...field
        //       }
        //     }
        //   }
        // }

        // get(lookupTable, `${pathPrefix}.childField.${questionId}`)

        // trick to detect arrays?
        // if (fieldOptions._genFieldComponent === FieldArray && has(fieldOptions, '_genChildren')) {
        //   console.log('*&*&&*&*&*&*&&* caught array', field);
        // }

        // if (fieldOptions._genTraverseChildren) {
        //   const childFields = fieldOptions._genTraverseChildren({
        //     iterator: ({field}) => field
        //   });
        // }
        if (field.childFields) {
          buildLookupTable(
            {
              ...options,
              fields: field.childFields
            },
            table
          );
        }

        if (fieldOptions._genChildren) {
          buildLookupTable(
            {
              ...options,
              fields: fieldOptions._genChildren
            },
            table
          );
        }
      }
    });
  }
  return table;
};

export const hasFieldDefaultValue = (options: GetDefaultValueOptions) => {
  const {field} = options;
  const fieldOptions = resolveFieldOptions(options);
  return has(field, 'defaultValue') || has(fieldOptions, '_genDefaultValue');
};

export const getFieldDefaultValue = (options: GetDefaultValueOptions) => {
  const {field} = options;
  const fieldOptions = resolveFieldOptions(options);
  return has(field, 'defaultValue')
    ? field.defaultValue
    : has(fieldOptions, '_genDefaultValue') ? fieldOptions._genDefaultValue : '';
};

export const getDefaultValues = (options: GetDefaultValuesOptions) => {
  options = {
    // defaultValues: {},
    ...options,
    data: options.initialValues || {},
    defaultValues: defaultsDeep(options.defaultValues, options.initialValues)
  };

  const {fields} = options;
  let {defaultValues} = options;

  fields.map((field) => getDefaultValue({...options, field}));
  return defaultValues;
};

const getDefaultValue = (options: GetDefaultValueOptions) => {
  options = {
    // defaultValues: {},
    ...options,
    data: options.initialValues || {},
    defaultValues: defaultsDeep(options.defaultValues, options.initialValues),
    fieldOptions: resolveFieldOptions(options)
  };

  // TODO does this need conditionalVisible (with lookupTable) support?
  // should a field still generate a defaultValue even if not visible? I think so...
  // so most likely don't need that

  const {field, fieldOptions} = options;
  let {defaultValues} = options;

  if (!isNil(fieldOptions)) {
    // only parse valid field types

    if (!isFieldHidden(options)) {
      // don't generate defaultValues for _genHidden fields.
      const fieldPath = getFieldPath(options);
      if (has(field, 'questionId') && !has(defaultValues, fieldPath)) {
        // might need isNilOrEmpty(get(defaultValues, fieldPath)) in the future
        if (hasFieldDefaultValue(options)) {
          set(defaultValues, fieldPath, getFieldDefaultValue(options));
        }
      }
    }

    mapFieldChildren({...options, fieldOptions: null}, getDefaultValue);
  }
  return defaultValues;
};

// used to merge existing and null/undefined paths together correctly
export const mergePaths = (...paths: Array<string>) => paths.filter((v) => v).join('.');
