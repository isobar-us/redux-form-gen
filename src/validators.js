// @flow
import isNil from 'lodash/isNil';
import trim from 'lodash/trim';
import isEmpty from 'lodash/isEmpty';

import get from 'lodash/get';
import set from 'lodash/set';
import has from 'lodash/has';
import omit from 'lodash/omit';
import values from 'lodash/values';
import isPlainObject from 'lodash/isPlainObject';
import isArray from 'lodash/isArray';
// import merge from 'lodash/merge';

import {evalCond, evalCondValid} from './conditionalUtils';
import {getFieldOptions} from './defaultFieldTypes';
import {mergePaths} from './utils';

export const isNilOrEmpty = (value: mixed) =>
  // $FlowFixMe
  isNil(value) || isEmpty(trim(value)) || ((isPlainObject(value) || isArray(value)) && isEmpty(value));

// ####################################################
// # Valid
// ####################################################

import type {FieldsType} from './types';

import type {
  SectionValidOptions,
  FieldValidOptions,
  SectionFilledOptions,
  FieldFilledOptions
} from './validators.types';

// options = {fields, data, lookupTable, customFieldTypes, parentQuestionId, errors = {}}
export const getSectionErrors = (options: SectionValidOptions) => {
  options = {
    // fields
    // customFieldTypes
    // data
    errors: {},
    ...options
  };

  const {fields, parent} = options;
  let {errors} = options;

  const parentQuestionId = parent && parent.questionId;

  fields.map((field) => getFieldErrors({...options, parentQuestionId, field}));
  return errors;
};

export const INVALID_MESSAGE = 'Invalid Field';
export const REQUIRED_MESSAGE = 'Required Field';

// options = {field, data, lookupTable, customFieldTypes, parentQuestionId, errors = {}}
export const getFieldErrors = (options: FieldValidOptions) => {
  options = {
    // field
    // customFieldTypes
    // data
    errors: {},
    messages: {
      requiredMessage: REQUIRED_MESSAGE,
      invalidMessage: INVALID_MESSAGE,
      ...options.messages
    },
    ...options
  };

  const {field, customFieldTypes, pathPrefix, data, parentQuestionId, messages} = options;
  let {errors} = options;

  const requiredMessage = has(field, 'requiredMessage') ? field.requiredMessage : messages.requiredMessage;
  const invalidMessage = has(field, 'invalidMessage') ? field.invalidMessage : messages.invalidMessage;

  const fieldOptions = getFieldOptions({field, customFieldTypes});

  // only parse valid field types
  if (!isNil(fieldOptions)) {
    const visible = has(fieldOptions, '_genHidden') ? !fieldOptions._genHidden : true;

    if (has(field, 'conditionalVisible')) {
      if (
        evalCond({
          ...options,
          cond: field.conditionalVisible,
          ...(parentQuestionId && {valueKey: parentQuestionId})
        })
      ) {
        getFieldErrors({...options, field: omit(field, 'conditionalVisible')});
      }
    } else if (visible) {
      const disabled =
        (has(field, 'disabled') && field.disabled) ||
        (has(fieldOptions, 'disabled') && fieldOptions.disabled) ||
        (has(field, 'conditionalDisabled') &&
          evalCond({
            ...options,
            cond: field.conditionalDisabled,
            ...(parentQuestionId && {valueKey: parentQuestionId})
          }));
      const required =
        (!disabled && (has(field, 'required') && field.required)) ||
        (has(field, 'conditionalRequired') &&
          evalCond({
            ...options,
            cond: field.conditionalRequired,
            ...(parentQuestionId && {valueKey: parentQuestionId})
          }));

      if (has(fieldOptions, 'name')) {
        // TODO doesn't work for <Fields names={...} />
        const path = mergePaths(pathPrefix, fieldOptions.name);
        const value = get(data, path);

        if (
          required === true &&
          (has(fieldOptions, '_genIsFilled') ? !fieldOptions._genIsFilled(options) : isNilOrEmpty(value))
        ) {
          set(errors, path, requiredMessage);
        } else if (has(fieldOptions, '_genIsValid') && !fieldOptions._genIsValid(options)) {
          set(errors, path, invalidMessage);
        } else if (has(fieldOptions, '_genSectionErrors')) {
          fieldOptions._genSectionErrors(options);
        } else if (
          has(field, 'conditionalValid') &&
          !evalCondValid({
            ...options,
            cond: field.conditionalValid,
            ...(field.questionId && {valueKey: field.questionId})
          })
        ) {
          set(errors, path, invalidMessage);
        }
      } else {
        // // if there is no questionId, and the is _genIsValid,
        // // then expect the result of _genIsValid to be an errors object, and merge the result
        // // this is needed to support multifields
        // if (has(fieldOptions, '_genIsValid')) {
        //   const fieldErrors = fieldOptions._genIsValid({
        //     ...options,
        //     messages: {
        //       invalid: invalidMessage,
        //       required: requiredMessage
        //     }
        //   });
        //   if (!isEmpty(fieldErrors)) {
        //     merge(errors, fieldErrors);
        //   }
        // }
      }
      if (has(field, 'childFields') && Array.isArray(field.childFields)) {
        getSectionErrors({...options, parent: field, fields: field.childFields});
      }

      if (fieldOptions._genTraverseChildren) {
        fieldOptions._genTraverseChildren({
          ...options,
          iterator: getFieldErrors
        });
      } else if (has(fieldOptions, '_genChildren') && Array.isArray(fieldOptions._genChildren)) {
        getSectionErrors({...options, parent: field, fields: fieldOptions._genChildren});
      }
    }
  }
  return errors;
};

// ####################################################
// # Filled
// ####################################################

// options = {field, data, lookupTable, customFieldTypes, parentQuestionId}
export const isFieldFilled = (options: FieldFilledOptions) => {
  const {field, parentQuestionId} = options;
  let fieldFilled = true;

  if (has(field, 'conditionalVisible')) {
    if (
      evalCond({
        ...options,
        cond: field.conditionalVisible,
        ...(parentQuestionId && {valueKey: parentQuestionId})
      })
    ) {
      fieldFilled = fieldFilled && isFieldFilled({...options, field: omit(field, 'conditionalVisible')});
    }
  } else {
    const required =
      (has(field, 'required') && field.required) ||
      (has(field, 'conditionalRequired') &&
        evalCond({
          ...options,
          cond: field.conditionalRequired,
          ...(parentQuestionId && {valueKey: parentQuestionId})
        }));

    const fieldOptions = getFieldOptions(options);

    let conditionalValid = field.conditionalValid || {};

    // TODO: move away from this pattern, try out the _genIsFilled pattern
    if (required) {
      if (has(fieldOptions, '_genIsFilled')) {
        fieldFilled = fieldFilled && fieldOptions._genIsFilled({...options});
      } else {
        let conditionalRequiredValid = {
          filled: true
        };
        fieldFilled =
          fieldFilled &&
          evalCondValid({
            ...options,
            cond: conditionalRequiredValid,
            ...(field.questionId && {valueKey: field.questionId})
          });
      }
    }

    // console.log('hit standard', field, conditionalValid);

    if (!isEmpty(conditionalValid)) {
      fieldFilled =
        fieldFilled &&
        evalCondValid({
          ...options,
          cond: conditionalValid,
          ...(field.questionId && {valueKey: field.questionId})
        });
    }

    // console.log('is field filled', fieldFilled);

    // if (required && has(field, 'conditionalValid')) {
    //   fieldFilled = fieldFilled && evalConditionalValid(field.conditionalValid, data, valueKey: field.questionId, lookupTable, {shallow: true});
    // }

    // TODO: not sure if we need this exactly, might push into _genOperators.filled ? similar to _genOperators.valid? or is that just condValid?
    // if (has(fieldOptions, '_genIsFilled')) {
    //   fieldFilled = fieldFilled && fieldOptions._genIsFilled({...options, required});
    // }

    if (has(fieldOptions, '_genTraverseChildren')) {
      fieldFilled =
        fieldFilled &&
        fieldOptions
          ._genTraverseChildren({
            ...options,
            iterator: ({field, ...iteratorOptions}) => isFieldFilled({field, ...iteratorOptions})
          })
          .reduce((result, fieldResult) => result && fieldResult, true);
    } else if (has(fieldOptions, '_genChildren') && Array.isArray(fieldOptions._genChildren)) {
      fieldFilled = fieldFilled && isSectionFilled({...options, fields: fieldOptions._genChildren, parent: field});
    }

    // trick to detect arrays?
    // if (fieldOptions._genFieldComponent === FieldArray && has(fieldOptions, '_genChildren')) {
    //   console.log('*&*&&*&*&*&*&&* caught array', field);
    // }

    if (has(field, 'childFields') && Array.isArray(field.childFields)) {
      fieldFilled = fieldFilled && isSectionFilled({...options, fields: field.childFields, parent: field});
    }

    // if (has(field, 'childFields') && Array.isArray(field.childFields)) {
    //   fieldFilled = fieldFilled && isSectionFilled({data, fields: field.childFields, lookupTable, parent: field});
    // }
  }
  return fieldFilled;
};

// returns true if a section has any fields that have been filled out, false otherwise
// lookupTable is required if you have fields that are conditionalValid.
export const isSectionFilled = ({data, fields, lookupTable = {}, parent, ...options}: SectionFilledOptions) => {
  const parentQuestionId = parent && parent.questionId;
  return fields.reduce((valid, field) => {
    return valid && isFieldFilled({...options, field, data, lookupTable, parentQuestionId});
  }, true);
};

// ####################################################
// # Empty
// ####################################################

// TODO needs rewrite

// returns true if a section has any fields that have been filled out, false otherwise
export const isSectionEmpty = (data: Object, fields: FieldsType) => {
  return fields.reduce((empty, field) => {
    let fieldEmpty = true;

    fieldEmpty =
      has(field, 'questionId') && has(data, field.questionId) ? isNilOrEmpty(get(data, field.questionId)) : true;
    if (field.childFields) {
      fieldEmpty = fieldEmpty && isSectionEmpty(data, field.childFields);
    } else if (field.columns) {
      fieldEmpty = fieldEmpty && field.columns.reduce((colEmpty, col) => colEmpty && isSectionEmpty(data, col), true);
    } else if (field.type === 'sectionArray' || field.type === 'array') {
      // TODO detect arrays with _genTraverseChildren instead
      let sectionArray = get(data, field.questionId);
      if (Array.isArray(sectionArray)) {
        fieldEmpty =
          fieldEmpty &&
          sectionArray.reduce(
            (sectionValid, sectionValues) => sectionValid && isSectionEmpty({...data, ...sectionValues}, field.fields),
            true
          );
      }
    } else if (field.fields) {
      if (!Array.isArray(field.fields)) {
        fieldEmpty = fieldEmpty && isSectionEmpty(data, values(field.fields));
      }
    }

    return empty && fieldEmpty;
  }, true);
};
