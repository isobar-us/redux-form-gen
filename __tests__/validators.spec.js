import {buildLookupTable} from '../src/utils';
import exampleFieldTypes from '../stories/customFieldTypes';
import {isSectionFilled, getSectionErrors, REQUIRED_MESSAGE, INVALID_MESSAGE, isNilOrEmpty} from '../src/validators';
import get from 'lodash/get';

describe('isSectionFilled()', () => {
  const fields = [
    {
      questionId: 'text_id',
      type: 'text',
      required: true
    },
    {
      questionId: 'array_id',
      type: 'array',
      item: {
        type: 'arrayItem',
        childFields: [
          {
            type: 'text',
            questionId: 'array_item_text',
            required: true
          },
          {
            type: 'radio',
            questionId: 'array_item_radio',
            options: [{label: 'Yes', value: 'yes'}, {label: 'Nes', value: 'no'}]
          },
          {
            type: 'array',
            questionId: 'array_nested',
            item: {
              type: 'arrayItem',
              childFields: [
                {
                  type: 'text',
                  questionId: 'array_nested_item_text',
                  required: true
                }
              ]
            }
          }
        ]
      }
    }
  ];

  it('should calculate filled true', () => {
    const data = {
      text_id: 'something',
      array_id: [
        {
          array_item_text: 'something',
          array_nested: [
            {
              array_nested_item_text: 'one more'
            }
          ]
        }
      ]
    };
    const lookupTable = buildLookupTable({fields, customFieldTypes: exampleFieldTypes});

    const filled = isSectionFilled({data, fields, customFieldTypes: exampleFieldTypes, lookupTable});
    expect(filled).toBe(true);
  });

  it('should calculate filled false', () => {
    const data = {
      text_id: 'something',
      array_id: [
        {
          array_nested: [
            {
              array_nested_item_text: 'one more'
            }
          ]
        }
      ]
    };
    const lookupTable = buildLookupTable({fields, customFieldTypes: exampleFieldTypes});

    const filled = isSectionFilled({data, fields, customFieldTypes: exampleFieldTypes, lookupTable});
    expect(filled).toBe(false);
  });
});

describe('getSectionErrors', () => {
  const fields = [
    {
      type: 'text',
      questionId: 'foo',
      required: true,
      conditionalValid: {
        pobox: true
      },
      childFields: [
        {
          type: 'text',
          questionId: 'child_foo',
          required: true,
          childFields: [
            {
              type: 'text',
              questionId: 'child_child_foo',
              required: true,
              conditionalValid: {
                questionId: 'child_foo',
                value: 'cond'
              }
            }
          ]
        }
      ]
    },
    {
      type: 'array',
      questionId: 'baz',
      item: {
        type: 'arrayItem',
        childFields: [
          {
            type: 'text',
            questionId: 'disbar',
            required: true,
            disabled: true
          },
          {
            type: 'text',
            questionId: 'condisbar',
            required: true,
            conditionalDisabled: {
              questionId: 'child_foo',
              value: 'cond'
            }
          },
          {
            type: 'text',
            questionId: 'bar',
            conditionalValid: {
              greaterThan: 10
            },
            required: true
          },
          {
            type: 'array',
            questionId: 'nested_baz',
            conditionalValid: {
              length: 2
            },
            item: {
              type: 'arrayItem',
              childFields: [
                {
                  type: 'text',
                  questionId: 'nested_foo',
                  conditionalValid: {
                    regex: '[^!]hello'
                  },
                  required: true
                }
              ]
            }
          }
        ]
      }
    }
  ];

  it('should message for required fields', () => {
    const data = {
      foo: 'Po Box 1234',
      child_foo: 'cond',
      child_child_foo: 'asdf',
      baz: [
        {
          nested_baz: [
            {
              nested_foo: 'ahello'
            },
            {
              nested_foo: 'bye'
            }
          ]
        }
      ]
    };

    const lookupTable = buildLookupTable({fields, customFieldTypes: exampleFieldTypes});
    const errors = getSectionErrors({
      fields,
      data,
      customFieldTypes: exampleFieldTypes,
      lookupTable
    });

    expect(errors).toEqual({
      // child_foo: REQUIRED_MESSAGE,
      baz: [
        {
          bar: REQUIRED_MESSAGE,
          nested_baz: [
            undefined, // eslint-disable-line no-undefined
            {
              nested_foo: INVALID_MESSAGE
            }
          ]
        }
      ]
    });

    const customRequiredMessage = 'Custom Required Message';
    const customInvalidMessage = 'Custom Invalid Message';

    const customErrors = getSectionErrors({
      fields,
      data,
      customFieldTypes: exampleFieldTypes,
      lookupTable,
      messages: {
        requiredMessage: customRequiredMessage,
        invalidMessage: customInvalidMessage
      }
    });

    expect(customErrors).toEqual({
      // child_foo: REQUIRED_MESSAGE,
      baz: [
        {
          bar: customRequiredMessage,
          nested_baz: [
            undefined, // eslint-disable-line no-undefined
            {
              nested_foo: customInvalidMessage
            }
          ]
        }
      ]
    });
  });

  it('should message for invalid fields', () => {
    const data = {
      foo: 'something',
      child_foo: 'what',
      child_child_foo: 'asdf',
      baz: [
        {
          bar: '2',
          nested_baz: [{}, {}]
        }
      ]
    };

    const lookupTable = buildLookupTable({fields, customFieldTypes: exampleFieldTypes});
    const errors = getSectionErrors({
      fields,
      data,
      customFieldTypes: exampleFieldTypes,
      lookupTable
    });

    expect(errors).toEqual({
      foo: INVALID_MESSAGE,
      child_child_foo: INVALID_MESSAGE,
      baz: [
        {
          condisbar: REQUIRED_MESSAGE,
          bar: INVALID_MESSAGE,
          nested_baz: [
            {
              nested_foo: REQUIRED_MESSAGE
            },
            {
              nested_foo: REQUIRED_MESSAGE
            }
          ]
        }
      ]
    });
  });

  describe('should use _genIsFilled for custom fields', () => {
    const customFieldTypes = {
      checkbox: ({field}) => ({
        _genIsFilled: ({field, data}) => get(data, field.questionId) === true,
        name: field.questionId
      })
    };

    const checkFields = [
      {
        type: 'checkbox',
        required: true,
        questionId: 'check'
      }
    ];

    const lookupTable = buildLookupTable({fields: checkFields, customFieldTypes});

    it('should show errors', () => {
      const data = {
        check: false
      };
      const errorsFalse = getSectionErrors({
        fields: checkFields,
        data: data,
        customFieldTypes,
        lookupTable
      });

      expect(errorsFalse).toEqual({
        check: REQUIRED_MESSAGE
      });
    });

    it('should show no errors', () => {
      const data = {
        check: true
      };

      const errors = getSectionErrors({
        fields: checkFields,
        data: data,
        customFieldTypes,
        lookupTable
      });

      expect(errors).toEqual({});
    });
  });

  describe('should respect conditionalVisible', () => {
    const visFields = [
      {
        type: 'text',
        questionId: 'condText'
      },
      {
        type: 'text',
        questionId: 'visText',
        required: true,
        conditionalVisible: {
          questionId: 'condText',
          value: 'something'
        }
      }
    ];

    const lookupTable = buildLookupTable({fields: visFields});

    it('should show errors', () => {
      const data = {
        condText: 'something'
      };
      const errorsFalse = getSectionErrors({
        fields: visFields,
        data: data,
        lookupTable
      });

      expect(errorsFalse).toEqual({
        visText: REQUIRED_MESSAGE
      });
    });

    it('should show no errors', () => {
      const data = {};

      const errors = getSectionErrors({
        fields: visFields,
        data: data,
        lookupTable
      });

      expect(errors).toEqual({});
    });
  });

  describe('isNilOrEmpty', () => {
    // true
    it('should return true for no param', () => {
      expect(isNilOrEmpty()).toBe(true);
    });
    it('should return true for null', () => {
      expect(isNilOrEmpty(null)).toBe(true);
    });
    it('should return true for undefined', () => {
      expect(isNilOrEmpty(undefined)).toBe(true); // eslint-disable-line no-undefined
    });
    it('should return true for {}', () => {
      expect(isNilOrEmpty({})).toBe(true);
    });
    it('should return true for []', () => {
      expect(isNilOrEmpty([])).toBe(true);
    });
    it('should return true for empty string', () => {
      expect(isNilOrEmpty('')).toBe(true);
    });
    it('should return true for string with spaces', () => {
      expect(isNilOrEmpty('   ')).toBe(true);
    });

    // false
    it('should return false for "lol"', () => {
      expect(isNilOrEmpty('lol')).toBe(false);
    });
    it('should return false for ["lol"]', () => {
      expect(isNilOrEmpty(['lol'])).toBe(false);
    });
    it('should return false for {"lol": false}', () => {
      expect(isNilOrEmpty({lol: false})).toBe(false);
    });
    it('should return false for number', () => {
      expect(isNilOrEmpty(123)).toBe(false);
    });
    it('should return false for zero', () => {
      expect(isNilOrEmpty(0)).toBe(false);
    });
    it('should return false for new Number()', () => {
      expect(isNilOrEmpty(new Number(123))).toBe(false); // eslint-disable-line no-new-wrappers
    });
    it('should return false for boolean', () => {
      expect(isNilOrEmpty(false)).toBe(false);
    });
    it('should return false for new Boolean()', () => {
      expect(isNilOrEmpty(new Boolean(false))).toBe(false); // eslint-disable-line no-new-wrappers
    });
    it('should return false for new Date()', () => {
      expect(isNilOrEmpty(new Date())).toBe(false);
    });
  });
});
