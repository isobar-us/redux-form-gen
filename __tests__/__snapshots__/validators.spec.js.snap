// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSectionErrors onSetError callback should call onSetError for each error 1`] = `
Array [
  Array [
    Object {
      "message": "Required Field",
      "path": "requiredText",
      "type": "required",
    },
  ],
  Array [
    Object {
      "message": "Invalid Field",
      "path": "invalidText",
      "type": "invalid",
    },
  ],
]
`;
